### PR TITLE
[DO NOT MERGE] Faster JSONata evaluation by switching from generators to async/await

### PR DESCRIPTION
**Pull request by tlakomy**
This PR is meant to be merged into the main JSONata repo, but I'd like to get your final feedback before I create it.

Description:

This PR is greatly inspired by and builds upon @avaidyam's work in https://github.com/jsonata-js/jsonata/pull/508

Apart from allowing non-blocking/asynchronous functions - `fetch()`, `fs.readFile` etc. to run in parallel (as noted by @avaidyam's), there are additional benefits of switching to async/await - performance benefits.

I've included two test cases in `performance` directory which contain two example JSONata expressions that take a non-trivial amount of time given an example array with 256 elements (see items.json):

```
{ 
  "items": items.{ 
    "foo": "bar", 
    "count": $count($$.items[text != "" and text != "foo"].row) 
  } 
}
```

Link: https://try.jsonata.org/sPNirb3df (it actually causes `Expression evaluation timeout: Check for infinite loop` in JSONata Exercises) 

and

```
{
  "items": items#$i.{
    "foo": "bar",
    "label": $$.items[$i].label & " " & $$.items[text!="" and text!="test"][$i].text
  }
}
```
Link: https://try.jsonata.org/K4_Qr1Hof

I've used those two cases as a benchmark, all numbers were measured on my Apple M1 Max (I acknowledge this is not an average machine, so your results may vary):

**`master` branch:**
```
Group: performance
      ✓ case000.json: {"items": items.{"foo": "bar","count": $count($$.items[text!="" and text!="foo"].row)}} (1418ms)
      ✓ case001.json: {"items": items#$i.{"foo": "bar","label": $$.items[$i].label & " " & $$.items[text!="" and text!="test"][$i].text}} (1706ms)
```

**`async/await` version:**
```
Group: performance
      ✓ case000.json: {"items": items.{"foo": "bar","count": $count($$.items[text!="" and text!="foo"].row)}} (241ms)
      ✓ case001.json: {"items": items#$i.{"foo": "bar","label": $$.items[$i].label & " " & $$.items[text!="" and text!="test"][$i].text}} (304ms)
```

`case000.json` went down from ~1400ms to ~240ms and `case001.json` from ~1700ms to ~300ms

I've also measured how long it takes to run the entire test suite, results below:
**`master` branch:**
`npm t  20.83s user 1.09s system 138% cpu 15.776 total`

**`async/await` version:**
`npm t  8.14s user 0.62s system 128% cpu 6.846 total`

--- 

In addition to those changes this PR introduces `nyc` package in favour of istanbul following this comment: https://github.com/gotwarlost/istanbul/issues/904#issuecomment-453436157

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -107,9 +107,9 @@ const utils = (() => {
     }
 
     // istanbul ignore next
-    var $Symbol = typeof Symbol === "function" ? Symbol : {};
+    const iteratorSymbol = (typeof Symbol === "function" ? Symbol : {}).iterator || "@@iterator";
     // istanbul ignore next
-    var iteratorSymbol = $Symbol.iterator || "@@iterator";
+    const asyncIteratorSymbol = (typeof Symbol === "function" ? Symbol : {}).asyncIterator || "@@asyncIterator";
```
Comment (asennikov) [inline src/utils.js:112]: it seems like they are using `var` everywhere, do we care about that?

File: test/implementation-tests.js [hunk 2]
```diff
@@ -318,14 +315,14 @@ describe("Tests that bind Javascript functions", () => {
...
-         * @constructor
-         */
+          * `DOMException` is not available in our testing environment. Additionally, we can't
+          * just import the `domexception` module since it doesn't work on Node.js v4, which
+          * we still support. So, here's a fake skeleton implementation which has the relevant
+          * qualities we need to reproduce the bug, most importantly a read-only `message`
+          * property
+          * @param {string} message - Error message
+          * @constructor
+          */
```
Comment (asennikov) [inline test/implementation-tests.js:325]: I would avoid this diff if possible

File: test/implementation-tests.js [hunk 3]
```diff
@@ -823,38 +846,37 @@ describe("Test that yield platform specific results", () => {
...
- * @param {Number} timeout - max time in ms
- * @param {Number} maxDepth - max stack depth
- */
+  * Protect the process/browser from a runnaway expression
+  * i.e. Infinite loop (tail recursion), or excessive stack growth
+  *
+  * @param {Object} expr - expression to protect
+  * @param {Number} timeout - max time in ms
+  * @param {Number} maxDepth - max stack depth
+  */
```
Comment (asennikov) [inline test/implementation-tests.js:879]: another avoidable diff

File: test/implementation-tests.js [hunk 4]
```diff
@@ -864,7 +886,7 @@ function timeboxExpression(expr, timeout, maxDepth) {
             // stack too deep
             throw {
                 message:
-                    "Stack overflow error: Check for non-terminating recursive function.  Consider rewriting as tail-recursive.",
+                     "Stack overflow error: Check for non-terminating recursive function.  Consider rewriting as tail-recursive.",
```
Comment (asennikov) [inline test/implementation-tests.js:889]: here too

File: src/jsonata.js [hunk 5]
```diff
@@ -1566,9 +1577,9 @@ var jsonata = (function() {
         if(expr.thunk === true) {
             procedure.thunk = true;
         }
-        procedure.apply = function*(self, args) {
-            return yield * apply(procedure, args, input, self.environment);
-        };
+        procedure.apply = async function(self, args) {
+            return await apply(procedure, args, input, !!self ? self.environment : environment);
+        }
```
[the remark this answers is not in the document]
Comment (tlakomy) [reply src/jsonata.js:1582]: Paging @mthenw (j/k)

**Comment by avaidyam** [comment]
This is awesome! Thanks for continuing the work I started, and feel free to reuse any of my code if it proves useful as well!

**Comment by WojciechMatuszewskiStedi** [comment]
:shipit:

**Comment by bahrmichael** [comment]
Closing this since you added those changes to the base repository, and I updated our fork over the last days.

**Pull request closed without merging**